File: app/src/main/java/com/wb/logistics/network/rx/RxWrapperCallAdapter.java
package com.wb.logistics.network.rx;

import androidx.annotation.NonNull;

import com.wb.logistics.network.exceptions.ErrorResolutionStrategy;

import java.lang.reflect.Type;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Single;
import retrofit2.Call;
import retrofit2.CallAdapter;

public class RxWrapperCallAdapter<R> implements CallAdapter<R, Object> {

    @NonNull
    private final CallAdapter<R, Object> wrapped;
    @NonNull
    private final ErrorResolutionStrategy errorResolutionStrategy;

    RxWrapperCallAdapter(@NonNull CallAdapter<R, Object> wrapped,
                         @NonNull ErrorResolutionStrategy errorResolutionStrategy) {
        this.wrapped = wrapped;
        this.errorResolutionStrategy = errorResolutionStrategy;
    }

    @Override
    public Type responseType() {
        return wrapped.responseType();
    }

    @Override
    public Object adapt(@NonNull Call<R> call) {
        Object adaptedCall = wrapped.adapt(call);
        if (adaptedCall instanceof Observable) {
            return errorResolutionStrategy.apply((Observable) adaptedCall);
        } else if (adaptedCall instanceof Single) {
            return errorResolutionStrategy.apply((Single) adaptedCall);
        } else if (adaptedCall instanceof Completable) {
            return errorResolutionStrategy.apply((Completable) adaptedCall);
        }
        throw new UnsupportedOperationException("Only Observable or Single or Completable supported");
    }

}

File: app/src/main/java/ru/wb/go/network/rx/RxWrapperCallAdapter.kt
package ru.wb.go.network.rx

import io.reactivex.Completable
import io.reactivex.Maybe
import io.reactivex.Observable
import io.reactivex.Single
import retrofit2.Call
import retrofit2.CallAdapter
import ru.wb.go.network.exceptions.ErrorResolutionStrategy
import java.lang.reflect.Type

class RxWrapperCallAdapter<R> internal constructor(
    private val callAdapter: CallAdapter<R, *>,
    private val errorResolutionStrategy: ErrorResolutionStrategy
) : CallAdapter<R, Any> {

    override fun responseType(): Type {
        return callAdapter.responseType()
    }

    override fun adapt(call: Call<R>): Any {
        return when (val adaptedCall = callAdapter.adapt(call)) {
            is Observable<*> -> errorResolutionStrategy.apply(adaptedCall)
            is Single<*> -> errorResolutionStrategy.apply(adaptedCall)
            is Completable -> errorResolutionStrategy.apply(adaptedCall)
            is Maybe<*> -> errorResolutionStrategy.apply(adaptedCall)
            else -> throw UnsupportedOperationException("Only Observable or Single or Completable or Maybe supported")
        }
    }

}